base/cvd/cuttlefish/host/libs/web/caching_build_api.cpp

```diff
@@ -0,0 +1,141 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "host/libs/web/caching_build_api.h"
+
+#include <chrono>
+#include <filesystem>
+#include <memory>
+#include <string>
+
+#include <android-base/strings.h>
+#include <fmt/format.h>
+
+#include "common/libs/utils/files.h"
+#include "common/libs/utils/result.h"
+#include "host/libs/web/android_build_api.h"
+#include "host/libs/web/credential_source.h"
+#include "host/libs/web/http_client/http_client.h"
+
+namespace cuttlefish {
+namespace {
+
+struct CachingPaths {
+  std::string build_cache;
+  std::string target_artifact;
+  std::string cache_artifact;
+  std::string target_backup_artifact;
+  std::string cache_backup_artifact;
+};
+
+bool EnsureCacheDirectory(const std::string& cache_base_path) {
+  Result<void> result = EnsureDirectoryExists(cache_base_path);
+  if (!result.ok()) {
+    LOG(WARNING) << "Failed to create cache directory \"" << cache_base_path
+                 << "\" with error: " << result.error().FormatForEnv();
+    LOG(WARNING) << "Caching disabled";
+    return false;
+  }
+  return true;
+}
+
+CachingPaths ConstructCachePaths(const std::string& cache_base,
+                                 const Build& build,
+                                 const std::string& target_directory,
+                                 const std::string& artifact,
+                                 const std::string& backup_artifact = "") {
+  const auto [id, target] = GetBuildIdAndTarget(build);
+  auto result = CachingPaths{
+      .build_cache = fmt::format("{}/{}/{}", cache_base, id, target),
+      .target_artifact = ConstructTargetFilepath(target_directory, artifact),
+  };
+  result.cache_artifact = ConstructTargetFilepath(result.build_cache, artifact);
+  if (!backup_artifact.empty()) {
+    result.target_backup_artifact =
+        ConstructTargetFilepath(target_directory, backup_artifact);
+    result.cache_backup_artifact =
+        ConstructTargetFilepath(result.build_cache, backup_artifact);
+  }
+  return result;
+}
+
+}  // namespace
+
+CachingBuildApi::CachingBuildApi(
+    std::unique_ptr<HttpClient> http_client,
+    std::unique_ptr<HttpClient> inner_http_client,
+    std::unique_ptr<CredentialSource> credential_source, std::string api_key,
+    const std::chrono::seconds retry_period, std::string api_base_url,
+    const bool enable_caching, const std::string cache_base_path)
+    : BuildApi(std::move(http_client), std::move(inner_http_client),
+               std::move(credential_source), std::move(api_key), retry_period,
+               std::move(api_base_url)),
+      enable_caching_(enable_caching),
+      cache_base_path_(std::move(cache_base_path)) {
+  // the cache directory must exist for the CanCache check to be true
+  enable_caching_ = enable_caching_ && EnsureCacheDirectory(cache_base_path_);
```

Comment:
An alternative to accepting `enable_caching` as an argument and member variable here is to create a public function
```
std::unique_ptr<BuildApi> CreateBuildApi(...);
```
that internally creates either kind.

Reply:
Good idea, I like that better than the internal member variable.  Done.